Added I2C bus masks

tridge opened the pull request:
This adds a system for I2C device bus masks, allowing for compasses on any external I2C bus
This is needed for fmuv5
@lucasdemarchi, it needs a bit more work to fill in the right mask values for each Linux port

ChristopherOlson commented:
This will get the I2C#4 port to work on the Pixhack v5?

OXINARF commented on libraries/AP_Notify/Display.cpp:
```diff
@@ -326,7 +323,7 @@ bool Display::init(void)
     // initialise driver
     for(uint8_t i=0; i<8 && _driver == nullptr; i++) {
-        if (! (I2C_BUS_PROBE_MASK & (1<<i))) {
+        if (! (hal.i2c_mgr->get_i2c_bus_mask_external() & (1<<i))) {
```
This doesn't work on the Cube because both buses are external.

tridge replied:
it does work in exactly the way it worked before (with the both_external boolean)
I didn't change the mask for the cube as it would change the bus order, so would break peoples configs

OXINARF replied:
Where is the both_external boolean? I think you are confusing Compass with Display...

tridge replied:
ahh, yes, sorry, I thought you meant compass. You're right that this will break display on i2c0 on cube

tridge replied:
fixed

tridge commented:
@ChristopherOlson i2c4 does actually work on the PixHackV5 in master, but the SDA and SCL pins are swapped in hardware. cuav is aware of the error.

tridge commented:
@OXINARF the aim of this PR is to get the API into master while not changing behavior (apart from now supporting buses beyond two for boards that have them)
So I didn't add probing on all buses for compasses that only probed on specific buses at the moment as that would be a behaviour change

tridge commented:
for board dependent masks, we could indeed have a way to change the mask depending on BRD_TYPE, but we'd need to think about how it may break existing configs

OXINARF commented:
> for board dependent masks, we could indeed have a way to change the mask depending on BRD_TYPE, but we'd need to think about how it may break existing configs

@tridge Well, if it's a huge code improvement (and in my opinion it is), then, if we need to break user configs, we should (it would be different if it wasn't detected, but it is).

lucasdemarchi commented on libraries/AP_HAL/I2CDevice.h:
```diff
@@ -89,6 +89,28 @@ class I2CDeviceManager {
+    /*
+      get mask of bus numbers for all configured internal I2C buses
+     */
+    virtual uint8_t get_i2c_bus_mask_internal(void) const { return 0x01; }
```
you are already in a I2CDeviceManager class... this could just be called "get_bus_mask_*()".

lucasdemarchi replied:
Can we make this uint32_t? At least then it would be compatible with most Linux boards we support.  It's not compatible today anyway because the numbers can change... that's why I added support for traversing sysfs to get a bus by the controller name

tridge replied:
we loop over all values checking a mask, using a uint32_t would take too long
we could make it a list instead?

tridge replied:
right now it looks like all the linux boards use i2c bus numbers 4 or less. Is that likely to change?

lucasdemarchi replied:
> we loop over all values checking a mask, using a uint32_t would take too long
> we could make it a list instead?

32 is still very small for a "do it on init only". If it's available on ARM, then __builtin_ffs would be an option to "return the next bit set"

> right now it looks like all the linux boards use i2c bus numbers 4 or less. Is that likely to change?

MS5611 on minlure is on bus 10

tridge replied:
yes, 32 is fine. I was being an idiot and thinking we'd loop over 2^32 :-)

lucasdemarchi commented on libraries/AP_HAL/board/linux.h:
```diff
@@ -406,3 +406,16 @@
+
+// if bus masks are not setup above then use these defaults
+#ifndef HAL_LINUX_I2C_BUS_MASK
+#define HAL_LINUX_I2C_BUS_MASK 0x1f
```
see above

tridge commented:
@lucasdemarchi I've changed to 32 bit and renamed the functions. I've also defaulted HAL_Linux to 16 buses.
I'm not sure if that is the best way to handle it though. What do you think?

lucasdemarchi commented:
> I'm not sure if that is the best way to handle it though. What do you think?

In future we will need to better differentiate internal vs external buses. Right now it's wrong: you should make internal I2C bus to be 0xFFFF as well.
